### PR TITLE
Add support for file connectors (File system and FTP)

### PR DESCRIPTION
**Pull request by pubudu91**
This PR will add transport support for the following connectors:
- FTP Server Connector
- FTP Client Connector
- File System Server Connector
- File Server Connector

This PR contains the commits of the following PRs: #299 , #301 , #314

### CODE REVIEW

File: file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/ThreadPoolFactory.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * under the License.
+ */
+package org.wso2.carbon.transport.filesystem.connector.server.util;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ *
+ */
```
Comment (shafreenAnfar) [inline file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/ThreadPoolFactory.java]: Requires proper documentation

File: connector-framework/src/main/java/org/wso2/carbon/connector/framework/server/polling/PollingJob.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+public class PollingJob implements Job {
+
+    private static final Logger log = LoggerFactory.getLogger(PollingJob.class);
+    @Override
+    public void execute(JobExecutionContext jobExecutionContext) throws JobExecutionException {
+        SchedulerContext schedulerContext = null;
+        try {
+            schedulerContext = jobExecutionContext.getScheduler().getContext();
+        } catch (SchedulerException e1) {
+            log.debug("Exception occurred while getting scheduler context");
```
Comment (ramith) [inline connector-framework/src/main/java/org/wso2/carbon/connector/framework/server/polling/PollingJob.java]: out of curiosity, logging with out the exception e1 is intentional ?

File: file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/FileSystemConsumer.java [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+     * @param children The array containing child elements of a folder
+     */
+    private void directoryHandler(FileObject[] children) throws FileSystemServerConnectorException {
+        // Sort the files according to given properties
+        String strSortParam = fileProperties.get(Constants.FILE_SORT_PARAM);
+
+        if (strSortParam != null && !"NONE".equals(strSortParam)) {
+            log.debug("Starting to sort the files in folder: " + FileTransportUtils.maskURLPassword(fileURI));
+
+            String strSortOrder = fileProperties.get(Constants.FILE_SORT_ORDER);
```
Comment (shafreenAnfar) [inline file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/FileSystemConsumer.java]: Sorting should be done in different method.

This breaks rule that a give method should only do one thing. This method has by products.

File: connector-framework/src/main/java/org/wso2/carbon/connector/framework/server/polling/PollingTaskRunner.java [hunk 4]
```diff
@@ -17,54 +17,71 @@
...
-            connector.poll();
-        } catch (Exception e) {
-            log.error("Error executing the polling cycle for " +
-                    "server connector ID: " + connector.getId(), e);
+            scheduler = new StdSchedulerFactory().getScheduler();
+            scheduler.getContext().put("connector", connector);
+            scheduler.start();
+            scheduler.scheduleJob(job, trigger);
+        } catch (SchedulerException e) {
+            log.error("Exception occurred while scheduling job");
```
Comment (ramith) [inline connector-framework/src/main/java/org/wso2/carbon/connector/framework/server/polling/PollingTaskRunner.java]: is it intentional not to log the 'e' ?

File: connector-framework/src/main/java/org/wso2/carbon/connector/framework/server/polling/PollingTaskRunner.java [hunk 5]
```diff
@@ -17,54 +17,71 @@
...
-     * Exit the running while loop and terminate the thread
+     * Exit the running while loop and terminate the thread.
      */
     protected void terminate() {
-        if (future != null) {
-            future.cancel(true);
+        try {
+            scheduler.shutdown();
+        } catch (SchedulerException e) {
+            log.error("Exception occurred when shutting down scheduler");
```
Comment (ramith) [inline connector-framework/src/main/java/org/wso2/carbon/connector/framework/server/polling/PollingTaskRunner.java]: same as above comment.

File: file/org.wso2.carbon.transport.file/src/main/java/org/wso2/carbon/transport/file/connector/sender/VFSClientConnector.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+/**
+ * VFSClientConnector.
+ */
```
Comment (shafreenAnfar) [inline file/org.wso2.carbon.transport.file/src/main/java/org/wso2/carbon/transport/file/connector/sender/VFSClientConnector.java]: Shall we add a proper doc.

File: file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/FileTransportUtils.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+    private static final Logger log = LoggerFactory.getLogger(FileSystemConsumer.class);
+    private static List<String> processing = new ArrayList<>();
+
+    private static final Pattern URL_PATTERN = Pattern.compile("[a-z]+://.*");
+    private static final Pattern PASSWORD_PATTERN = Pattern.compile(":(?:[^/]+)@");
+
+    /**
+     *
+     * @param options
```
Comment (shafreenAnfar) [inline file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/FileTransportUtils.java]: Need proper documentation

File: file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/FileTransportUtils.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+                                                             + options.get(Constants.SCHEME) + " and option: "
+                                                             + Constants.FILE_TYPE, e);
+            }
+        }
+        return opts;
+    }
+
+    /**
+     *
+     * @param url
```
Comment (shafreenAnfar) [inline file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/FileTransportUtils.java]: Need documentation.

File: file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/FileTransportUtils.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            String maskUrl = pwdMatcher.replaceFirst("\":***@\"");
+            return maskUrl;
+        } else {
+            return url;
+        }
+    }
+
+    /**
+     *
+     * @param fileType
```
Comment (shafreenAnfar) [inline file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/FileTransportUtils.java]: Need proper documentation.

File: file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/ThreadPoolFactory.java (hunk 1 above)
Comment (pubudu91) [reply file/org.wso2.carbon.transport.file-system/src/main/java/org/wso2/carbon/transport/filesystem/connector/server/util/ThreadPoolFactory.java]: Ack.

File: connector-framework/src/main/java/org/wso2/carbon/connector/framework/server/polling/PollingJob.java (hunk 2 above)
Comment (pubudu91) [reply connector-framework/src/main/java/org/wso2/carbon/connector/framework/server/polling/PollingJob.java]: Will fix this.

File: file/org.wso2.carbon.transport.file/src/main/java/org/wso2/carbon/transport/file/connector/sender/VFSClientConnector.java (hunk 6 above)
Comment (pubudu91) [reply file/org.wso2.carbon.transport.file/src/main/java/org/wso2/carbon/transport/file/connector/sender/VFSClientConnector.java]: Ack.

**Review by shafreenAnfar — APPROVED** [review]

**Pull request merged**